content/automate/ManageIQ/Cloud/VM/Provisioning/StateMachines/Methods.class/__methods__/amazon_customizerequest.rb

```diff
@@ -2,99 +2,166 @@
+                def ws_values
+                  @ws_values ||= prov.options.fetch(:ws_values, {})
+                end
+
+                def process_customization
+                  set_options
+                  set_customization_template
+                end
+
+                def set_options
+                  if prov.get_option(:instance_type).nil? && ws_values.key?(:instance_type)
+                    provider.flavors.each do |flavor|
+                      if flavor.name.downcase == ws_values[:instance_type].downcase
+                        prov.set_option(:instance_type, [flavor.id, "#{flavor.name}':'#{flavor.description}"])
+                        log(:info, "Provisioning object updated {:instance_type => #{prov.get_option(:instance_type).inspect}}")
+                      end
+                    end
+                  end
 
-# Get provisioning object
-prov = $evm.root["miq_provision"]
+                  if prov.get_option(:guest_access_key_pair).nil? && ws_values.key?(:guest_access_key_pair)
+                    provider.key_pairs.each do |keypair|
+                      if keypair.name == ws_values[:guest_access_key_pair]
+                        prov.set_option(:guest_access_key_pair, [keypair.id, keypair.name])
+                        log(:info, "Provisioning object updated {:guest_access_key_pair => #{prov.get_option(:guest_access_key_pair).inspect}}")
+                      end
+                    end
+                  end
 
-log(:info, "Provision:<#{prov.id}> Request:<#{prov.miq_provision_request.id}> Type:<#{prov.type}>")
+                  if prov.get_option(:security_groups).blank? && ws_values.key?(:security_groups)
+                    provider.security_groups.each do |securitygroup|
+                      if securitygroup.name == ws_values[:security_groups]
+                        prov.set_option(:security_groups, [securitygroup.id])
+                        log(:info, "Provisioning object updated {:security_groups => #{prov.get_option(:security_groups).inspect}}")
+                      end
+                    end
+                  end
+                end
 
-template = prov.vm_template
-provider = template.ext_management_system
-product  = template.operating_system['product_name'].downcase rescue nil
-log(:info, "Template: #{template.name} Provider: #{provider.name} Vendor: #{template.vendor} Product: #{product}")
+                def set_customization_template
+                  if prov.get_option(:customization_template_id).present?
+                    log(:info, "Customization Template selected from dialog ID: %{id} Script: %{script}" % {
+                      :id     => prov.get_option(:customization_template_id).inspect,
+                      :script => prov.get_option(:customization_template_script).inspect
+                    })
+                    return
+                  end
+
+                  customization_template_search_by_function       = "#{prov.type}_#{prov.get_tags[:function]}" rescue nil
```

Comment:
The `prov` method will raise an error if it cannot return the object, so we can just remove the `rescue nil` at the end.